### PR TITLE
Legacy C++ gradients for PartitionedCall and StatefulPartitionedCall

### PR DESCRIPTION
**Pull request by rnett**
This adds legacy C++ (`cc/gradients`) gradients for `PartitionedCall` and `StatefulPartitionedCall`, using `SymbolicGradient` as per https://github.com/tensorflow/tensorflow/pull/46115#issuecomment-797685426.

### CODE REVIEW

**Review by allenlavoie — APPROVED** [review]

**Comment by allenlavoie** [comment]
gradients_functional_grad_test is missing BUILD dependencies for its includes. Please add `//tensorflow/cc:ops`

**Review by allenlavoie — APPROVED** [review]

File: tensorflow/cc/gradients/functional_grad_test.cc [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#include "tensorflow/cc/framework/gradient_checker.h"
+#include "tensorflow/cc/framework/testutil.h"
+#include "tensorflow/cc/gradients/grad_testutil.h"
+#include "tensorflow/core/framework/tensor_testutil.h"
+#include "tensorflow/core/lib/core/status_test_util.h"
+
+namespace tensorflow {
+namespace {
+
+using namespace ops;  // NOLINT(build/namespaces)
```
Comment (allenlavoie) [inline tensorflow/cc/gradients/functional_grad_test.cc:31]: Clang tidy doesn't like this ([neither does the style guide](https://google.github.io/styleguide/cppguide.html#Namespaces)). Should this test just be in the ops namespace too?

Comment (rnett) [reply tensorflow/cc/gradients/functional_grad_test.cc:31]: Seems fine, I just copied it from the other tests.

Should I do this (and add the build dependency) for #46115 too?

Comment (allenlavoie) [reply tensorflow/cc/gradients/functional_grad_test.cc]: Ugh. Yeah, looks like that's why that one hasn't submitted itself either yet. The suggestions from the automated tool for that one:

  Add //tensorflow/cc:gradients
  Add //tensorflow/cc:ops
  Add //tensorflow/cc:client_session

(to gradients_resource_variable_grad_test)

Very annoying when the internal CI is stricter than the checks we have here.

**Review by allenlavoie — APPROVED** [review]

**Pull request merged**